src/tunnel.cc

```diff
@@ -444,6 +444,9 @@ TunnelStateData::retryOrBail(const char *context)
     // sendNewError() and sendSavedErrorOr(), used in "error detected" cases.
     if (!savedError)
         saveError(new ErrorState(ERR_CANNOT_FORWARD, Http::scInternalServerError, request.getRaw(), al));
+
+    al->updateError(Error(savedError->type, savedError->detail));
```

Comment:
Does _this_ line fix the `Also fixes/logs %err_code value for ERR_RELAY_REMOTE transactions.` part of the PR description? It is unclear, though, why it is limited only by ERR_RELAY_REMOTE cases. I suspect that that since ErrorState() constructor does not get ALE for ERR_RELAY_REMOTE, it is not updated somewhere automatically and   we need to do it manually, but I could not find the code responsible for that.

In FwdState we call al->updateError() from FwdState::updateAleWithFinalError(). It probably should be called in all error cases, including Tunneler problems resulting in ERR_RELAY_REMOTE.

Reply:
> missing your underlying concern.

The PR (and its description) are about fixing missing err_detail. It also states that one missing err_code ERR_RELAY_REMOTE was fixed, it does not say that some other missing codes might have been fixed as well. If my understanding is correct, than we should probably note there that the err_code fix in tunnel.cc comprises not only ERR_RELAY_REMOTE transactions, but other error transactions as well.

Reply:
> The PR (and its description) are about fixing missing err_detail. It also states that one missing err_code ERR_RELAY_REMOTE was fixed, it does not say that some other missing codes might have been fixed as well. 

Correct. The last bit is always a possibility. I do not think we have to explicitly mention it, especially when we are not sure whether any other error code could have been missing. As you know, ERR_RELAY_REMOTE could have been a special case because its corresponding constructor was missing ALE.

> we should probably note there that the err_code fix in tunnel.cc comprises not only ERR_RELAY_REMOTE transactions, but other error transactions as well.

To say something more specific than "at least ERR_RELAY_REMOTE", we would need to have at least strong suspicion that "other error transactions" are affected. We did not have it -- all other errors in our tests relayed %err_code fine without that tunnel.cc fix.

I have now tested the theory that the missing constructor ALE was responsible for the missing ERR_RELAY_REMOTE and proved it wrong: Commenting out updateError() in TunnelStateData::retryOrBail() restores the missing ERR_RELAY_REMOTE bug despite the fixed ErrorState constructor.

Further investigation uncovered the true cause of the ERR_RELAY_REMOTE bug: ErrorState::BuildHttpReply() bypasses request->detailError() calls (at the end of that method) if error reply is already known, which is the case for ERR_RELAY_REMOTE but not for any other error (because only ERR_RELAY_REMOTE sets ErrorState::response_ which effectively disables BuildHttpReply()).

My current plan is to split this PR into two, isolating ERR_RELAY_REMOTE-related/inspired fixes into a separate PR.

Reply:
> Further investigation uncovered the true cause of the ERR_RELAY_REMOTE bug: ErrorState::BuildHttpReply() bypasses request->detailError() calls (at the end of that method) if the error response message is already known, which is the case for ERR_RELAY_REMOTE but not for any other error (because only ERR_RELAY_REMOTE sets ErrorState::response_ which effectively disables BuildHttpReply()).

Fixed in branch commit d6c71f6.